crates/fm/src/lib.rs

```diff
@@ -10,28 +10,42 @@ pub use file_map::{File, FileId, FileMap};
 
 use std::{
     collections::HashMap,
-    path::{Path, PathBuf},
+    path::{Component, Path, PathBuf},
 };
 
 pub const FILE_EXTENSION: &str = "nr";
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 struct VirtualPath(PathBuf);
 
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct FileManager {
+    root: PathBuf,
     file_map: file_map::FileMap,
     id_to_path: HashMap<FileId, VirtualPath>,
     path_to_id: HashMap<VirtualPath, FileId>,
 }
 
 impl FileManager {
-    // XXX: Maybe use a AsRef<Path> here, for API ergonomics
-    pub fn add_file(&mut self, path_to_file: &Path) -> Option<FileId> {
+    pub fn new(root: &Path) -> Self {
+        Self {
+            root: normalize_path(root),
+            file_map: Default::default(),
+            id_to_path: Default::default(),
+            path_to_id: Default::default(),
+        }
+    }
+
+    pub fn add_file(&mut self, file_name: &Path) -> Option<FileId> {
```

Comment:
I think we can rewrite this function to avoid normalizing the stdlib paths entirely.

We currently only call `normalize_path` outside of the `else` branch due to needing to maintain the lifetime of `resolved_path: &Path`. If we switch `resolved_path` to be a `PathBuf` then we can remove this.

```rust
    pub fn add_file(&mut self, file_name: &Path) -> Option<FileId> {
        // Handle both relative file paths and std/lib virtual paths.
        let resolved_path: PathBuf =
            if file_name.starts_with("std/") || file_name.starts_with("std\\") {
                // Special case for stdlib where we want to read specifically the `std/` relative path
                // TODO: The stdlib path should probably be an absolute path rooted in something people would never create
                file_name.to_path_buf()
            } else {
                 normalize_path(&self.root.join(file_name))
            };

        // Check that the resolved path already exists in the file map, if it is, we return it.
        let path_to_file = virtualize_path(&resolved_path);
        if let Some(file_id) = self.path_to_id.get(&path_to_file) {
            return Some(*file_id);
        }

        // Otherwise we add the file
        let source = file_reader::read_file_to_string(&resolved_path).ok()?;
        let file_id = self.file_map.add_file(resolved_path.into(), source);
        self.register_path(file_id, path_to_file);
        Some(file_id)
    }
```

Reply:
Can likely even use a `Cow<Path>` to avoid allocating the pathbuf for the std case entirely